File: internal/sdk/python/mindclade_internal_sdk/testing.py
"""Hermetic transports for repository-internal SDK tests and consumer fakes."""

from __future__ import annotations

import hashlib
from collections.abc import AsyncIterator, Callable, Iterable, Iterator
from dataclasses import dataclass
from datetime import UTC, datetime

from google.protobuf.message import Message
from google.protobuf.timestamp_pb2 import Timestamp
from mindclade.artifact.v1 import artifact_reference_pb2
from mindclade.common.v1 import event_envelope_pb2
from mindclade.internal.artifact.v1 import artifact_service_pb2
from mindclade.internal.job.v1 import job_service_pb2
from mindclade.job.v1 import job_pb2, job_requested_pb2

from .transport import Metadata

SyncUnaryHandler = Callable[[Message, float, Metadata], Message]
SyncStreamHandler = Callable[[Message, float, Metadata], Iterable[Message]]
AsyncUnaryHandler = Callable[[Message, float, Metadata], Message]
AsyncStreamHandler = Callable[[Message, float, Metadata], AsyncIterator[Message]]


def get_job_request_name(request: Message) -> str:
    """Inspect a generated GetJob request without exposing its module to a consumer test."""

    if not isinstance(request, job_service_pb2.GetJobRequest):
        raise TypeError("expected the SDK's generated GetJob request")
    return request.name


def artifact_download_request_digest(request: Message) -> str:
    """Inspect a generated artifact download request behind the SDK fake seam."""

    if not isinstance(request, artifact_service_pb2.DownloadArtifactRequest):
        raise TypeError("expected the SDK's generated DownloadArtifact request")
    return request.digest


@dataclass(frozen=True, slots=True)
class RecordedCall:
    """Payload-free call record safe to inspect in tests."""

    method: str
    timeout: float
    metadata_keys: tuple[str, ...]


class FakeSyncTransport:
    def __init__(self) -> None:
        self.unary_handlers: dict[str, SyncUnaryHandler] = {}
        self.response_metadata: dict[str, Metadata] = {}
        self.stream_handlers: dict[str, SyncStreamHandler] = {}
        self.calls: list[RecordedCall] = []
        self.closed = False

    def unary_unary(
        self,
        method: str,
        request: Message,
        *,
        timeout: float,
        metadata: Metadata,
    ) -> Message:
        self.calls.append(RecordedCall(method, timeout, tuple(sorted(key for key, _ in metadata))))
        return self.unary_handlers[method](request, timeout, metadata)

    def unary_unary_with_metadata(
        self,
        method: str,
        request: Message,
        *,
        timeout: float,
        metadata: Metadata,
    ) -> tuple[Message, Metadata]:
        response = self.unary_unary(method, request, timeout=timeout, metadata=metadata)
        return response, self.response_metadata.get(method, ())

    def unary_stream(
        self,
        method: str,
        request: Message,
        *,
        timeout: float,
        metadata: Metadata,
    ) -> Iterator[Message]:
        self.calls.append(RecordedCall(method, timeout, tuple(sorted(key for key, _ in metadata))))
        yield from self.stream_handlers[method](request, timeout, metadata)

    def close(self) -> None:
        self.closed = True


class FakeAsyncTransport:
    def __init__(self) -> None:
        self.unary_handlers: dict[str, AsyncUnaryHandler] = {}
        self.response_metadata: dict[str, Metadata] = {}
        self.stream_handlers: dict[str, AsyncStreamHandler] = {}
        self.calls: list[RecordedCall] = []
        self.closed = False

    async def unary_unary(
        self,
        method: str,
        request: Message,
        *,
        timeout: float,
        metadata: Metadata,
    ) -> Message:
        self.calls.append(RecordedCall(method, timeout, tuple(sorted(key for key, _ in metadata))))
        return self.unary_handlers[method](request, timeout, metadata)

    async def unary_unary_with_metadata(
        self,
        method: str,
        request: Message,
        *,
        timeout: float,
        metadata: Metadata,
    ) -> tuple[Message, Metadata]:
        response = await self.unary_unary(method, request, timeout=timeout, metadata=metadata)
        return response, self.response_metadata.get(method, ())

    async def unary_stream(
        self,
        method: str,
        request: Message,
        *,
        timeout: float,
        metadata: Metadata,
    ) -> AsyncIterator[Message]:
        self.calls.append(RecordedCall(method, timeout, tuple(sorted(key for key, _ in metadata))))
        async for item in self.stream_handlers[method](request, timeout, metadata):
            yield item

    async def close(self) -> None:
        self.closed = True


def artifact_fixture(
    content: bytes, *, media_type: str = "application/json"
) -> artifact_reference_pb2.ArtifactRef:
    """Build a complete provider-neutral artifact identity for consumer tests."""

    return artifact_reference_pb2.ArtifactRef(
        digest="sha256:" + hashlib.sha256(content).hexdigest(),
        media_type=media_type,
        size_bytes=len(content),
    )


def job_response_fixture(
    configuration: artifact_reference_pb2.ArtifactRef,
    *,
    input_artifact: artifact_reference_pb2.ArtifactRef | None = None,
    tenant_id: str = "tenant-1",
    project_id: str = "project-1",
) -> job_service_pb2.GetJobResponse:
    """Build one valid durable job response behind the SDK fake boundary."""

    job = job_pb2.Job(
        job_id="jobs/job-1",
        operation_id="operations/op-1",
        tenant_id=tenant_id,
        project_id=project_id,
        state=job_pb2.JOB_STATE_RUNNING,
        resource_version=1,
        configuration=configuration,
        etag="job-etag-1",
    )
    if input_artifact is not None:
        job.input.CopyFrom(input_artifact)
    return job_service_pb2.GetJobResponse(job=job)


def artifact_download_fixture(
    artifact: artifact_reference_pb2.ArtifactRef,
    content: bytes,
    *,
    offset: int = 0,
    complete: bool = True,
) -> artifact_service_pb2.DownloadArtifactResponse:
    """Build one digest-bound download response for a consumer fake."""

    return artifact_service_pb2.DownloadArtifactResponse(
        artifact=artifact,
        offset=offset,
        data=content,
        chunk_digest="sha256:" + hashlib.sha256(content).hexdigest(),
        complete=complete,
    )


def job_requested_delivery_fixture(
    configuration_digest: str,
    *,
    tenant_id: str = "tenant-1",
    project_id: str = "project-1",
    event_version: int = 1,
    payload_digest: str | None = None,
) -> bytes:
    """Encode a deterministic immutable delivery for consumer tests."""

    event = job_requested_pb2.JobRequested(
        job_id="jobs/job-1",
        configuration_digest=configuration_digest,
    )
    payload = event.SerializeToString(deterministic=True)
    timestamp = Timestamp()
    timestamp.FromDatetime(datetime.now(UTC))
    envelope = event_envelope_pb2.EventEnvelope(
        event_id="events/event-1",
        event_type="mindclade.events.job.v1.JobRequested",
        event_version=event_version,
        occurred_at=timestamp,
        recorded_at=timestamp,
        tenant_id=tenant_id,
        project_id=project_id,
        aggregate_sequence=1,
        request_id="request-1",
        trace_id="trace-1",
        job_id=event.job_id,
        payload=payload,
        payload_digest=payload_digest or "sha256:" + hashlib.sha256(payload).hexdigest(),
        payload_content_type="application/x-protobuf; deterministic=true",
    )
    envelope.subject.resource_type = "job"
    envelope.subject.resource_id = "job-1"
    return envelope.SerializeToString(deterministic=True)

File: sdks/rust/src/testing.rs
//! Hermetic SDK-owned fakes for application-consumer tests.

use std::{
    collections::{HashMap, VecDeque},
    future::pending,
    sync::Mutex,
};

use mindclade_protocols::{
    artifact::v1::ArtifactRef,
    common::v1::{EventEnvelope, ResourceRef},
    internal::{
        artifact::v1::{DownloadArtifactRequest, DownloadArtifactResponse},
        job::v1::{GetJobRequest, GetJobResponse},
    },
    job::v1::{Job, JobRequested, JobState},
};
use prost::Message;
use prost_types::Timestamp;
use sha2::{Digest, Sha256};
use tonic::{Code, Request, Response, Status, codegen::async_trait, codegen::tokio_stream};

use crate::{ArtifactStream, JitterSource, RpcTransport};

#[must_use]
/// Builds a digest-consistent artifact reference for hermetic SDK consumer tests.
///
/// # Panics
///
/// Panics only when the fixture is larger than the protocol's signed 64-bit
/// size field, which cannot occur in a practical in-memory test fixture.
pub fn artifact_fixture(content: &[u8], media_type: &str) -> ArtifactRef {
    ArtifactRef {
        digest: sha256(content),
        media_type: media_type.to_owned(),
        size_bytes: i64::try_from(content.len()).expect("test artifact fits i64"),
        ..ArtifactRef::default()
    }
}

#[must_use]
pub fn job_fixture(
    tenant_id: &str,
    project_id: &str,
    configuration: ArtifactRef,
    input: Option<ArtifactRef>,
) -> Job {
    Job {
        job_id: "jobs/job-1".to_owned(),
        operation_id: "operations/operation-1".to_owned(),
        tenant_id: tenant_id.to_owned(),
        project_id: project_id.to_owned(),
        state: JobState::Running as i32,
        resource_version: 1,
        configuration: Some(configuration),
        input,
        etag: "job-etag-1".to_owned(),
        ..Job::default()
    }
}

#[must_use]
pub fn job_requested_delivery_fixture(
    configuration_digest: &str,
    tenant_id: &str,
    project_id: &str,
    event_version: u32,
    payload_digest_override: Option<&str>,
) -> Vec<u8> {
    let event = JobRequested {
        job_id: "jobs/job-1".to_owned(),
        configuration_digest: configuration_digest.to_owned(),
    };
    let payload = event.encode_to_vec();
    EventEnvelope {
        event_id: "events/event-1".to_owned(),
        event_type: "mindclade.events.job.v1.JobRequested".to_owned(),
        event_version,
        occurred_at: Some(Timestamp {
            seconds: 1,
            nanos: 0,
        }),
        tenant_id: tenant_id.to_owned(),
        trace_id: "trace-1".to_owned(),
        subject: Some(ResourceRef {
            resource_type: "job".to_owned(),
            resource_id: "job-1".to_owned(),
            ..ResourceRef::default()
        }),
        payload_digest: payload_digest_override.map_or_else(|| sha256(&payload), str::to_owned),
        payload,
        recorded_at: Some(Timestamp {
            seconds: 1,
            nanos: 0,
        }),
        project_id: project_id.to_owned(),
        aggregate_sequence: 1,
        request_id: "request-1".to_owned(),
        job_id: "jobs/job-1".to_owned(),
        payload_content_type: "application/x-protobuf; deterministic=true".to_owned(),
        ..EventEnvelope::default()
    }
    .encode_to_vec()
}

/// A content-addressed fake for `Jobs.Get` and `Artifacts.Download`.
pub struct ScriptedJobArtifactTransport {
    job: Job,
    artifacts: HashMap<String, Vec<u8>>,
    get_job_failure: Option<Code>,
    block_get_job: bool,
    requested_digests: Mutex<Vec<String>>,
}

impl ScriptedJobArtifactTransport {
    #[must_use]
    pub fn new(job: Job, artifacts: impl IntoIterator<Item = (ArtifactRef, Vec<u8>)>) -> Self {
        Self {
            job,
            artifacts: artifacts
                .into_iter()
                .map(|(artifact, content)| (artifact.digest, content))
                .collect(),
            get_job_failure: None,
            block_get_job: false,
            requested_digests: Mutex::new(Vec::new()),
        }
    }

    #[must_use]
    pub fn failing_get_job(mut self, code: Code) -> Self {
        self.get_job_failure = Some(code);
        self
    }

    #[must_use]
    pub fn blocking_get_job(mut self) -> Self {
        self.block_get_job = true;
        self
    }

    #[must_use]
    pub fn requested_digests(&self) -> Vec<String> {
        self.requested_digests
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner)
            .clone()
    }
}

#[async_trait]
impl RpcTransport for ScriptedJobArtifactTransport {
    async fn get_job(
        &self,
        request: Request<GetJobRequest>,
    ) -> Result<Response<GetJobResponse>, Status> {
        if self.block_get_job {
            pending::<()>().await;
        }
        if let Some(code) = self.get_job_failure {
            return Err(Status::new(code, "redacted fake failure"));
        }
        if request.get_ref().name != self.job.job_id {
            return Err(Status::not_found("unknown fake job"));
        }
        Ok(Response::new(GetJobResponse {
            job: Some(self.job.clone()),
        }))
    }

    async fn download_artifact(
        &self,
        request: Request<DownloadArtifactRequest>,
    ) -> Result<Response<ArtifactStream>, Status> {
        let digest = request.get_ref().digest.clone();
        self.requested_digests
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner)
            .push(digest.clone());
        let artifact = [&self.job.configuration, &self.job.input]
            .into_iter()
            .flatten()
            .find(|artifact| artifact.digest == digest)
            .cloned()
            .ok_or_else(|| Status::not_found("unknown fake artifact"))?;
        let content = self
            .artifacts
            .get(&digest)
            .cloned()
            .ok_or_else(|| Status::not_found("missing fake artifact content"))?;
        let response = DownloadArtifactResponse {
            artifact: Some(artifact),
            offset: 0,
            chunk_digest: sha256(&content),
            data: content,
            complete: true,
        };
        let stream: ArtifactStream = Box::pin(tokio_stream::iter([Ok(response)]));
        Ok(Response::new(stream))
    }
}

fn sha256(content: &[u8]) -> String {
    format!("sha256:{:x}", Sha256::digest(content))
}

/// Scripted retry jitter for hermetic tests.
///
/// Each entry is a fraction of the caller's full-jitter window, consumed in
/// order; the queue repeats its final entry once exhausted, so a test never
/// depends on how many delays the SDK happens to draw.
#[derive(Debug)]
pub struct ScriptedJitter {
    fractions: Mutex<VecDeque<f64>>,
    last: Mutex<f64>,
}

impl ScriptedJitter {
    /// Always returns the full capped exponential window.
    #[must_use]
    pub fn max() -> Self {
        Self::repeating(1.0)
    }

    /// Always returns a zero delay.
    #[must_use]
    pub fn zero() -> Self {
        Self::repeating(0.0)
    }

    /// Returns each fraction of the window in turn, then repeats the last one.
    /// Values outside `[0.0, 1.0]` are clamped rather than rejected so a test
    /// fixture can never produce a delay outside the SDK's own window.
    #[must_use]
    pub fn scripted(fractions: Vec<f64>) -> Self {
        let last = fractions.last().copied().unwrap_or(1.0);
        Self {
            fractions: Mutex::new(fractions.into()),
            last: Mutex::new(clamp_fraction(last)),
        }
    }

    fn repeating(fraction: f64) -> Self {
        Self {
            fractions: Mutex::new(VecDeque::new()),
            last: Mutex::new(clamp_fraction(fraction)),
        }
    }
}

impl JitterSource for ScriptedJitter {
    // The scripted fraction is clamped to `[0.0, 1.0]` and the product is
    // clamped to the caller's window, so the conversions below are bounded by
    // construction and a rounding difference on a very large window is
    // irrelevant to a test fixture.
    #[allow(
        clippy::cast_precision_loss,
        clippy::cast_possible_truncation,
        clippy::cast_sign_loss
    )]
    fn jitter_micros(&self, upper_bound_micros: u64) -> u64 {
        let mut queue = self
            .fractions
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner);
        let mut last = self
            .last
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner);
        if let Some(next) = queue.pop_front() {
            *last = clamp_fraction(next);
        }
        let scaled = *last * upper_bound_micros as f64;
        // `scaled` is finite and within `[0, upper_bound_micros]` by
        // construction, so the cast cannot saturate to an out-of-range value.
        (scaled.round() as u64).min(upper_bound_micros)
    }
}

fn clamp_fraction(value: f64) -> f64 {
    if value.is_nan() {
        return 0.0;
    }
    value.clamp(0.0, 1.0)
}
